apollo-runtime/src/commonMain/kotlin/com/apollographql/apollo3/internal/DeferredJsonMerger.kt

```diff
@@ -0,0 +1,112 @@
+package com.apollographql.apollo3.internal
+
+import com.apollographql.apollo3.annotations.ApolloInternal
+import com.apollographql.apollo3.api.json.BufferedSourceJsonReader
+import com.apollographql.apollo3.api.json.readAny
+import okio.BufferedSource
+
+/**
+ * Utility class for merging GraphQL JSON payloads received in multiple chunks when using the `@defer` directive.
+ *
+ * Each call to [merge] will merge the given chunk into the [merged] Map, and will also update the [mergedFragmentLabels] Set with the
+ * value of its `label` field (if present).
+ *
+ * The fields in `data` are merged into the node found in [merged] at `path` (for the first call to [merge], the payload is
+ * copied to [merged] as-is).
+ *
+ * Errors are merged by adding them to [merged]'s `errors` list field.
+ *
+ * Extensions are merged by adding them into [merged]'s `extensions` field - if several payloads have the same key in `extensions`,
+ * their values are stored as a list.
+ */
+@ApolloInternal
+class DeferredJsonMerger {
+  val merged: Map<String, Any?> = mutableMapOf()
+
+  private val _mergedFragmentLabels = mutableSetOf<String>()
+  val mergedFragmentLabels: Set<String> = _mergedFragmentLabels
+
+  fun merge(payload: BufferedSource): Map<String, Any?> {
+    val payloadMap = jsonToMap(payload)
+    if (merged.isEmpty()) {
+      // Initial payload, no merging needed
+      (merged as MutableMap) += payloadMap
+      return merged
+    }
+
+    mergeData(payloadMap)
+    mergeErrors(payloadMap)
+    mergeExtensions(payloadMap)
+
+    return merged
+  }
+
+  @Suppress("UNCHECKED_CAST")
+  private fun mergeData(payloadMap: Map<String, Any?>) {
+    val payloadData = payloadMap["data"] as Map<String, Any?>?
+    val payloadPath = payloadMap["path"] as List<Any>
+    val mergedData = merged["data"] as Map<String, Any?>
+
+    // payloadData can be null if there are errors
+    if (payloadData != null) {
+      val nodeToMergeInto = nodeAtPath(mergedData, payloadPath) as MutableMap<String, Any?>
+      nodeToMergeInto += payloadData
```

Comment:
This will overwrite merged fields if any, you need to deep merge. 

Example:

```graphql
query Query1 {
  computers {
    id
    screen {
       isColor
    }
    ...ComputerFields @defer
  }
}

fragment ComputerFields on Computer {
  # writing screen will lose `isColor`
  screen {
    resolution
  }
}```

Reply:
Good catch! 👍